File: chat_agent.py
#!/usr/bin/env python3
"""
Chat Agent - Agente de chat con respuestas preprogramadas y capacidad para agregar instrucciones
"""

import json
import os
from typing import Dict, List, Optional
from datetime import datetime


class ChatAgent:
    """
    Agente de chat que puede responder mensajes preprogramados y aceptar nuevas instrucciones.
    """
    
    def __init__(self, config_file: str = "config.json"):
        """
        Inicializa el agente de chat.
        
        Args:
            config_file: Ruta al archivo de configuración con respuestas preprogramadas
        """
        self.config_file = config_file
        self.responses = {}
        self.custom_instructions = []
        self.conversation_history = []
        self.load_config()
    
    def load_config(self):
        """Carga las respuestas preprogramadas desde el archivo de configuración."""
        if os.path.exists(self.config_file):
            try:
                with open(self.config_file, 'r', encoding='utf-8') as f:
                    data = json.load(f)
                    self.responses = data.get('responses', {})
                    self.custom_instructions = data.get('custom_instructions', [])
                    print(f"✓ Configuración cargada desde {self.config_file}")
            except Exception as e:
                print(f"✗ Error al cargar la configuración: {e}")
                self.initialize_default_config()
        else:
            print(f"⚠ Archivo de configuración no encontrado. Creando configuración por defecto...")
            self.initialize_default_config()
    
    def initialize_default_config(self):
        """Inicializa las respuestas preprogramadas por defecto."""
        self.responses = {
            "hola": "¡Hola! Soy el agente de chat de NBC Sport. ¿En qué puedo ayudarte?",
            "ayuda": "Puedo responder preguntas sobre deportes, eventos y horarios. También puedes agregar nuevas instrucciones usando el comando 'agregar instrucción'.",
            "deportes": "NBC Sport transmite una amplia variedad de deportes incluyendo fútbol, básquetbol, béisbol, hockey y más.",
            "horarios": "Los horarios de transmisión varían según el evento. Por favor, visita nuestro sitio web para información actualizada.",
            "contacto": "Puedes contactarnos a través de nuestro sitio web o redes sociales oficiales.",
            "adios": "¡Hasta luego! Que tengas un excelente día.",
            "gracias": "¡De nada! Estoy aquí para ayudarte.",
        }
        self.custom_instructions = []
        self.save_config()
    
    def save_config(self):
        """Guarda la configuración actual en el archivo."""
        try:
            data = {
                'responses': self.responses,
                'custom_instructions': self.custom_instructions,
                'last_updated': datetime.now().isoformat()
            }
            with open(self.config_file, 'w', encoding='utf-8') as f:
                json.dump(data, f, ensure_ascii=False, indent=2)
            print(f"✓ Configuración guardada en {self.config_file}")
        except Exception as e:
            print(f"✗ Error al guardar la configuración: {e}")
    
    def add_instruction(self, keyword: str, response: str):
        """
        Agrega una nueva instrucción/respuesta al agente.
        
        Args:
            keyword: Palabra clave que activará la respuesta
            response: Respuesta a dar cuando se detecte la palabra clave
        """
        keyword = keyword.lower().strip()
        self.responses[keyword] = response
        
        instruction_record = {
            'keyword': keyword,
            'response': response,
            'added_at': datetime.now().isoformat()
        }
        self.custom_instructions.append(instruction_record)
        
        self.save_config()
        print(f"✓ Nueva instrucción agregada: '{keyword}'")
    
    def find_response(self, message: str) -> Optional[str]:
        """
        Busca una respuesta apropiada para el mensaje dado.
        
        Args:
            message: Mensaje del usuario
            
        Returns:
            Respuesta encontrada o None si no hay coincidencia
        """
        message_lower = message.lower().strip()
        
        # Buscar coincidencia exacta
        if message_lower in self.responses:
            return self.responses[message_lower]
        
        # Buscar coincidencia parcial (priorizar palabras clave más largas)
        matches = []
        for keyword, response in self.responses.items():
            if keyword in message_lower:
                matches.append((len(keyword), keyword, response))
        
        # Retornar la coincidencia con la palabra clave más larga
        if matches:
            matches.sort(reverse=True)  # Ordenar por longitud descendente
            return matches[0][2]  # Retornar la respuesta
        
        return None
    
    def process_message(self, message: str) -> str:
        """
        Procesa un mensaje y genera una respuesta.
        
        Args:
            message: Mensaje del usuario
            
        Returns:
            Respuesta del agente
        """
        # Agregar mensaje al historial
        self.conversation_history.append({
            'timestamp': datetime.now().isoformat(),
            'user': message,
            'agent': None
        })
        
        # Buscar respuesta
        response = self.find_response(message)
        
        if response is None:
            response = "No tengo una respuesta preprogramada para eso. Puedes agregar una nueva instrucción o reformular tu pregunta."
        
        # Actualizar historial con la respuesta
        self.conversation_history[-1]['agent'] = response
        
        return response
    
    def list_instructions(self):
        """Lista todas las instrucciones/respuestas disponibles."""
        print("\n=== Instrucciones Preprogramadas ===")
        for i, (keyword, response) in enumerate(self.responses.items(), 1):
            print(f"{i}. '{keyword}' -> {response[:50]}{'...' if len(response) > 50 else ''}")
        
        if self.custom_instructions:
            print("\n=== Instrucciones Personalizadas ===")
            for i, instruction in enumerate(self.custom_instructions, 1):
                print(f"{i}. '{instruction['keyword']}' (Agregada: {instruction['added_at']})")
    
    def show_conversation_history(self, limit: int = 10):
        """
        Muestra el historial de conversación.
        
        Args:
            limit: Número máximo de mensajes a mostrar
        """
        print("\n=== Historial de Conversación ===")
        recent_history = self.conversation_history[-limit:]
        for entry in recent_history:
            print(f"\n[{entry['timestamp']}]")
            print(f"Usuario: {entry['user']}")
            print(f"Agente: {entry['agent']}")
    
    def clear_history(self):
        """Limpia el historial de conversación."""
        self.conversation_history = []
        print("✓ Historial de conversación limpiado")


def main():
    """Función principal para ejecutar el agente de chat."""
    print("=" * 60)
    print("        AGENTE DE CHAT - NBC SPORT")
    print("=" * 60)
    print("\nComandos especiales:")
    print("  /ayuda              - Muestra información de ayuda")
    print("  /listar             - Lista todas las instrucciones")
    print("  /agregar            - Agrega una nueva instrucción")
    print("  /historial          - Muestra el historial de conversación")
    print("  /limpiar            - Limpia el historial")
    print("  /salir              - Sale del programa")
    print("=" * 60)
    
    agent = ChatAgent()
    
    while True:
        try:
            print("\n")
            user_input = input("Tú: ").strip()
            
            if not user_input:
                continue
            
            # Comandos especiales
            if user_input.lower() == '/salir':
                print("\nAgente: ¡Hasta luego!")
                break
            
            elif user_input.lower() == '/ayuda':
                response = agent.find_response("ayuda")
                if response:
                    print(f"\nAgente: {response}")
                else:
                    print("\nAgente: Comando de ayuda disponible. Usa palabras clave como 'hola', 'deportes', 'horarios' o usa /listar para ver todas las opciones.")
            
            elif user_input.lower() == '/listar':
                agent.list_instructions()
            
            elif user_input.lower() == '/agregar':
                print("\n--- Agregar Nueva Instrucción ---")
                keyword = input("Palabra clave: ").strip()
                if not keyword:
                    print("✗ Palabra clave no puede estar vacía")
                    continue
                response_text = input("Respuesta: ").strip()
                if not response_text:
                    print("✗ Respuesta no puede estar vacía")
                    continue
                agent.add_instruction(keyword, response_text)
            
            elif user_input.lower() == '/historial':
                agent.show_conversation_history()
            
            elif user_input.lower() == '/limpiar':
                agent.clear_history()
            
            else:
                # Procesar mensaje normal
                response = agent.process_message(user_input)
                print(f"\nAgente: {response}")
        
        except KeyboardInterrupt:
            print("\n\nAgente: ¡Hasta luego!")
            break
        except Exception as e:
            print(f"\n✗ Error: {e}")


if __name__ == "__main__":
    main()

File: chat-agent.js
// Agente de Chat NBC Sport - Versión JavaScript
class ChatAgent {
    constructor() {
        this.responses = {
            "hola": "¡Hola! Soy el agente de chat de NBC Sport. ¿En qué puedo ayudarte?",
            "ayuda": "Puedo responder preguntas sobre deportes, eventos y horarios. También puedes usar comandos como /listar, /historial y /limpiar.",
            "deportes": "NBC Sport transmite una amplia variedad de deportes incluyendo fútbol, básquetbol, béisbol, hockey y más.",
            "horarios": "Los horarios de transmisión varían según el evento. Por favor, visita nuestro sitio web para información actualizada.",
            "contacto": "Puedes contactarnos a través de nuestro sitio web o redes sociales oficiales.",
            "adios": "¡Hasta luego! Que tengas un excelente día.",
            "gracias": "¡De nada! Estoy aquí para ayudarte.",
            "futbol": "NBC Sport transmite los mejores partidos de fútbol. Consulta nuestra programación para horarios específicos.",
            "basquetbol": "Disfruta de la mejor acción del básquetbol en NBC Sport, incluyendo la NBA y más.",
            "beisbol": "NBC Sport te trae toda la emoción del béisbol, incluyendo la MLB.",
            "noticias": "Visita nuestro sitio web para las últimas noticias deportivas y actualizaciones en tiempo real.",
            "streaming": "NBC Sport está disponible en múltiples plataformas de streaming. Consulta nuestro sitio web para más información.",
            "suscripcion": "Para información sobre suscripciones y paquetes, visita nuestra página web oficial o contacta a servicio al cliente."
        };
        
        this.conversationHistory = [];
        this.loadHistory();
    }

    // Cargar historial desde localStorage
    loadHistory() {
        const saved = localStorage.getItem('chatHistory');
        if (saved) {
            try {
                this.conversationHistory = JSON.parse(saved);
            } catch (e) {
                this.conversationHistory = [];
            }
        }
    }

    // Guardar historial en localStorage
    saveHistory() {
        try {
            localStorage.setItem('chatHistory', JSON.stringify(this.conversationHistory));
        } catch (e) {
            console.error('Error guardando historial:', e);
        }
    }

    // Buscar respuesta para un mensaje
    findResponse(message) {
        const lowerMessage = message.toLowerCase().trim();
        
        // Buscar coincidencia exacta
        if (this.responses[lowerMessage]) {
            return this.responses[lowerMessage];
        }
        
        // Buscar coincidencia parcial
        for (const [keyword, response] of Object.entries(this.responses)) {
            if (lowerMessage.includes(keyword)) {
                return response;
            }
        }
        
        return null;
    }

    // Procesar mensaje del usuario
    processMessage(message) {
        const timestamp = new Date().toISOString();
        
        // Agregar mensaje del usuario al historial
        this.conversationHistory.push({
            type: 'user',
            message: message,
            timestamp: timestamp
        });

        let response;

        // Procesar comandos especiales
        if (message.startsWith('/')) {
            response = this.handleCommand(message);
        } else {
            // Buscar respuesta
            response = this.findResponse(message);
            
            if (!response) {
                response = "Lo siento, no tengo información sobre eso. Intenta preguntar sobre deportes, horarios, streaming o usa /ayuda para ver los comandos disponibles.";
            }
        }

        // Agregar respuesta del agente al historial
        this.conversationHistory.push({
            type: 'agent',
            message: response,
            timestamp: new Date().toISOString()
        });

        this.saveHistory();
        return response;
    }

    // Manejar comandos especiales
    handleCommand(command) {
        const cmd = command.toLowerCase().trim();

        switch (cmd) {
            case '/ayuda':
                return `Comandos disponibles:\n\n` +
                       `• /ayuda - Muestra esta información\n` +
                       `• /listar - Lista todas las palabras clave disponibles\n` +
                       `• /historial - Muestra el historial de conversación\n` +
                       `• /limpiar - Limpia el historial de conversación\n\n` +
                       `También puedes preguntar sobre deportes, horarios, streaming y más.`;

            case '/listar':
                const keywords = Object.keys(this.responses);
                let list = '=== Instrucciones Disponibles ===\n\n';
                keywords.forEach((keyword, index) => {
                    const response = this.responses[keyword];
                    const preview = response.length > 50 ? response.substring(0, 50) + '...' : response;
                    list += `${index + 1}. '${keyword}' → ${preview}\n`;
                });
                return list;

            case '/historial':
                if (this.conversationHistory.length === 0) {
                    return 'No hay historial de conversación.';
                }
                
                let history = '=== Historial de Conversación ===\n\n';
                this.conversationHistory.forEach(entry => {
                    const date = new Date(entry.timestamp).toLocaleString('es-ES');
                    const prefix = entry.type === 'user' ? 'Usuario' : 'Agente';
                    history += `[${date}]\n${prefix}: ${entry.message}\n\n`;
                });
                return history;

            case '/limpiar':
                this.conversationHistory = [];
                this.saveHistory();
                return 'Historial de conversación limpiado.';

            default:
                return `Comando no reconocido: ${command}\nUsa /ayuda para ver los comandos disponibles.`;
        }
    }

    // Limpiar historial
    clearHistory() {
        this.conversationHistory = [];
        this.saveHistory();
    }
}

// Inicializar la aplicación cuando el DOM esté listo
document.addEventListener('DOMContentLoaded', () => {
    const chatAgent = new ChatAgent();
    const chatWindow = document.getElementById('chatWindow');
    const userInput = document.getElementById('userInput');
    const sendButton = document.getElementById('sendButton');

    // Función para agregar mensaje a la ventana de chat
    function addMessage(text, isUser = false) {
        const messageDiv = document.createElement('div');
        messageDiv.className = `message ${isUser ? 'user-message' : 'agent-message'}`;
        
        const contentDiv = document.createElement('div');
        contentDiv.className = 'message-content';
        
        const prefix = isUser ? '<strong>Tú:</strong> ' : '<strong>Agente:</strong> ';
        
        // Convertir saltos de línea y formato
        let formattedText = text.replace(/\n/g, '<br>');
        
        contentDiv.innerHTML = prefix + formattedText;
        messageDiv.appendChild(contentDiv);
        chatWindow.appendChild(messageDiv);
        
        // Scroll al final
        chatWindow.scrollTop = chatWindow.scrollHeight;
    }

    // Función para enviar mensaje
    function sendMessage() {
        const message = userInput.value.trim();
        
        if (message === '') {
            return;
        }

        // Mostrar mensaje del usuario
        addMessage(message, true);

        // Limpiar input
        userInput.value = '';

        // Procesar mensaje y obtener respuesta
        const response = chatAgent.processMessage(message);

        // Mostrar respuesta del agente con un pequeño delay
        setTimeout(() => {
            addMessage(response, false);
        }, 300);
    }

    // Event listeners
    sendButton.addEventListener('click', sendMessage);
    
    userInput.addEventListener('keypress', (e) => {
        if (e.key === 'Enter') {
            sendMessage();
        }
    });

    // Foco en el input al cargar
    userInput.focus();

    // Cargar historial previo si existe
    if (chatAgent.conversationHistory.length > 0) {
        const loadPrevious = confirm('¿Deseas cargar el historial de conversación anterior?');
        if (loadPrevious) {
            chatAgent.conversationHistory.forEach(entry => {
                addMessage(entry.message, entry.type === 'user');
            });
        } else {
            chatAgent.clearHistory();
        }
    }
});
